### PR TITLE
DSD-684: Created react chakra table

### PR DESCRIPTION
**Pull request by gonuguntla**
Fixes JIRA ticket [DSD-684](https://jira.nypl.org/browse/DSD-736)

## This PR does the following:
* Adds the Table components.

## How has this been tested?
* local build of DS Storybook

## Accessibility concerns or updates
- Right now implemented for only desktop. In future need to support for mobile.

### Checklist:

- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

### Front End Review:

- [x] View [the example in Storybook](https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa)

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/components/Table/Table.stories.mdx
Comment (EdwinGuzman) [inline src/components/Table/Table.stories.mdx]:
```
tableData: { control: false },
id: {control: false},
```

File: src/components/Table/Table.stories.mdx
Comment (EdwinGuzman) [inline src/components/Table/Table.stories.mdx]: Don't import this.

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx]: This is not right. You want to import `Table` DS component.

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx]: Remove this.

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx]:
```
<Table tableData={tableData} columnHeaders={columnHeaders} />
```

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Nitpick: remove spaces.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Remove the question mark `?` to make this a required prop.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Note: pass the logic check here for `useRowHeaders` when i is 0. And use 
```
_first: {
  fontWeight: "bold"
}
```
in `customTable.ts`.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Remove `Box`.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Remove the extra spaces.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]:
```
{columnHeaders.map((child, key) => {
  return <ChakraTh key={key} sx={customColors}> {child} </ChakraTh>;
```

Comment (EdwinGuzman) [reply src/components/Table/Table.tsx]: This is only if you want to. This is a shortcut but does the same thing without the `return` statement.

```
{columnHeaders.map((child) => (
          <ChakraTh sx={customColors}>{child}</ChakraTh>
))}
```

File: src/theme/components/customTable.ts
Comment (EdwinGuzman) [inline src/theme/components/customTable.ts]: Get a variable color.
```
color: columnHeadersTextColor ? columnHeadersTextColor : "ui.gray.medium",
```

File: src/theme/components/customTable.ts
Comment (EdwinGuzman) [inline src/theme/components/customTable.ts]: @bigfishdesign13 what is the theme equivalent for "sm" here?

File: src/theme/components/customTable.ts
Comment (EdwinGuzman) [inline src/theme/components/customTable.ts]: Updates these values.

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx:52]: Write at least one test checking that the desired text output is rendered.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Walked through the requested changes together. Let me know if you need more information.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/theme/components/customTable.ts
Comment (gonuguntla) [reply src/theme/components/customTable.ts]: @EdwinGuzman right now Marty told me use fontSize: "0". In future syntax is going to change

File: src/components/Table/Table.stories.mdx
Comment (EdwinGuzman) [inline src/components/Table/Table.stories.mdx:31]: You don't need this line here.

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx]: This test is the same as the test on line 32. The title for this test says "with children components". So you should try passing in children components for this example.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Minor: no `'` in "Displays".

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx:28]: This says that the default value is `false` but on line 44, it is set to `true`.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Set this to `= false` to match the description for this prop on line 26.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Nitpick and not a blocker: Would it make sense to rename "data" to "row"? And also to rename "tbody" to "column" so it makes it clear how it's looping through the data?

File: src/theme/components/customTable.ts
Comment (EdwinGuzman) [inline src/theme/components/customTable.ts]: Pass in `undefined` instead of `""`.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: What happens if `tableData` is _not_ a two dimensional array? What if this is passed `["one", "two", "three"]`?  Typescript won't allow this to pass but you should have a validation here for `tableData` and `console.warn` a message if the data is not two dimensional.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Some more minor changes but one of the tests should be updated and a warning message should be logged for data validation.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: Accordion to the spec document, "`Each header cell must have an appropriate scope attribute. Either scope="col" or scope="row"`". To achieve this, this `ChakraTh` component needs a `scope="col"` prop.

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: When `useRowHeaders` is true, besides making the first cell bold, it should be a `tr` instead of a `td`. According to the spec document, it also needs a `scope="row"`, 
`Each header cell must have an appropriate scope attribute. Either scope="col" or scope="row"`".

So, for example, when `useRowHeaders` is true, it should render something like
```
<tr>
  <th scope="row">first</th>
  <td>second</td>
  <td>third</td>
</tr>
```

This this as an example: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/th

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Besides the validation we talked over a video call, it just needs a minor update to make the first cell in a row a `th` when `useRowHeaders` is true. And, it also needs the appropriate `scope` prop for column and row.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/components/Table/Table.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.tsx]: When `useRowHeaders` is true, besides making the first cell bold, it should be a `tr` instead of a `td`. According to the spec document, it also needs a `scope="row"`, 
`Each header cell must have an appropriate scope attribute. Either scope="col" or scope="row"`".

So, for example, when `useRowHeaders` is true, it should render something like
```
<tr>
  <th scope="row">first</th>
  <td>second</td>
  <td>third</td>
</tr>
```

This this as an example: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/th

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
The update is good. The last thing is just adding the appropriate `scope="row"` or `scope="column"` as the spec doc mentions.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx]: Move this inside the second test, so inside "tableData prop is empty array" before line 55.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

File: src/components/Table/Table.test.tsx
Comment (EdwinGuzman) [inline src/components/Table/Table.test.tsx]: Update this to 
```it("logs a warning when the `tableData` prop is an empty array", async () => {```

**Review by EdwinGuzman — COMMENTED** [review]
Okay, just minor changes to one test and in the Storybook file, the table is off:
Screen Shot 2022-01-31 at 5 07 01 PM

it should be "Added" on top and then "Latest" below and they both point to `0.25.9`.

Everything else looks good for the first iteration of the `Table` component! I would like more tests and more examples in Storybook, however. Those can be done in the next iteration. This just needs one more look by @bigfishdesign13 .

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Review by EdwinGuzman — APPROVED** [review]
Approving. Wait for Marty's review before merging.

File: src/theme/components/customTable.ts
Comment (bigfishdesign13) [inline src/theme/components/customTable.ts]: The default color for the headers sholuld be `ui.black`.

File: src/theme/components/customTable.ts
Comment (bigfishdesign13) [inline src/theme/components/customTable.ts:6]: Add a style to set the letter spacing to 0.  You did that for `td`, but it also needs to be added here.
```
letterSpacing: "0",
```

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
There are some basic text style changes that need to be made (I added comments for those), but there are also some spacing changes that will require some more sophisticated fixes.

Update the padding for both `th` and `td` to the following:
```
paddingLeft: "0",
paddingRight: "m",
```

You will then need to add some conditional statements to the theme file to handle the left and right padding.  
- For both `th` and `td`, add `paddingLeft: "m"` to the first table cell in a row when `showRowDivider` is `true`.
- For both `th` and `td`, add `paddingRight: "0"` to the last table cell in a row when `showRowDivider` is `false`.

The styles for the `caption` element need to be updated to match the exmple that I just added to the Figma file.  Ultimately, the `caption` should be formatted to match the styles for `Heading.Secondary`.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Review by bigfishdesign13 — APPROVED** [review]
Looks good.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr846-wyo6rgmwzweh9ov4thhzaa3bj0zsyvwm.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61eae828ffbe18bf78263d58

**Pull request merged**
